compiler/generator_rust.c

```diff
@@ -0,0 +1,1376 @@
+
+#include <stdlib.h> /* for exit */
+#include <string.h> /* for strlen */
+#include <stdio.h> /* for fprintf etc */
+#include "header.h"
+
+/* prototypes */
+
+static void generate(struct generator * g, struct node * p);
+static void w(struct generator * g, const char * s);
+static void writef(struct generator * g, const char * s, struct node * p);
+
+
+enum special_labels {
+    x_return = -1
+};
+
+static int new_label(struct generator * g) {
+
+    return g->next_label++;
+}
+
+static struct str * vars_newname(struct generator * g) {
+
+    struct str * output;
+    g->var_number ++;
+    output = str_new();
+    str_append_string(output, "v_");
+    str_append_int(output, g->var_number);
+    return output;
+}
+
+
+/* Write routines for items from the syntax tree */
+
+static void write_varname(struct generator * g, struct name * p) {
+
+    switch (p->type) {
+        case t_external:
+            break;
+        default: {
+            int ch = "SbirxG"[p->type];
+            write_char(g, ch);
+            write_char(g, '_');
+            break;
+        }
+    }
+    str_append_b(g->outbuf, p->b);
+}
+
+static void write_varref(struct generator * g, struct name * p) {
+    write_string(g, "context.");
+    write_varname(g, p);
+}
+
+static void write_hexdigit(struct generator * g, int n) {
+
+    write_char(g, n < 10 ? n + '0' : n - 10 + 'A');
+}
+
+static void write_hex(struct generator * g, int ch) {
+
+    write_string(g, "\\u{");
+    {
+        int i;
+        for (i = 12; i >= 0; i -= 4) write_hexdigit(g, ch >> i & 0xf);
+    }
+    write_string(g, "}");
+}
+
+static void write_literal_string(struct generator * g, symbol * p) {
+
+    int i = 0;
+    write_string(g, "\"");
+    while (i < SIZE(p)) {
+        int ch;
+        i += get_utf8(p + i, &ch);
+        if (32 <= ch && ch <= 127) {
```

Comment:
Should be `ch < 127` (this flaw is in all the other backends except C, but I'm about to commit a fix for those).